### PR TITLE
Fix for #84 and documentation improvements

### PR DESCRIPTION
**Pull request by erasche**
Modify bioblend.galaxy.histories to implement TODOs, fix dataset URL by downloading from url provided by show_datasets() call.

This doesn't pass test cases on my desktop (but then again, reverting to the parent commit also didn't let it pass tests), so I'm going to let Travis test it and rely on those test results.

There is a small behaviour change here:

| file_path | use_default_filename | original output | new output
--- | --- | --- | ----
`str` (folder) | True | None | return path the file was saved to, rather than leaving the user guessing/calculating
`str` (file) | False| None | return the user requested path that the file was saved to
None | True or False | file contents | file contents

The other small behaviour change being that `download_dataset` in `bioblend.galaxy.histories` now returns the value of the call from `bioblend.galaxy.datasets` rather than `None`. 

Lastly, `HistoryClient.download_dataset` now ignores `history_id`, as that isn't actually needed to download a dataset.

### CODE REVIEW

File: bioblend/galaxy/histories/__init__.py [hunk 1]
```diff
@@ -16,6 +13,7 @@ class HistoryClient(Client):
 
     def __init__(self, galaxy_instance):
         self.module = 'histories'
+        self.galaxy_instance = galaxy_instance
```
Comment (nsoranzo) [inline bioblend/galaxy/histories/__init__.py]: This is not needed, ```galaxy_instance``` is already saved as ```self.gi``` in the superclass ```__init__()``` method.

Comment (erasche) [inline bioblend/galaxy/histories/__init__.py]: Ah, excellent. Will update presently.

File: bioblend/galaxy/histories/__init__.py [hunk 2]
```diff
@@ -261,26 +259,10 @@ def download_dataset(self, history_id, dataset_id, file_path,
...
-
-        req = urllib2.urlopen(url)
-        if use_default_filename:
-            file_local_path = os.path.join(file_path, meta['name'])
-        else:
-            file_local_path = file_path
-
-        with open(file_local_path, 'wb') as fp:
-            shutil.copyfileobj(req, fp)
+        dc = DatasetClient(self.galaxy_instance)
```
Comment (nsoranzo) [inline bioblend/galaxy/histories/__init__.py]: Use ```self.gi``` instead of ```self.galaxy_instance```.

Comment (erasche) [inline bioblend/galaxy/histories/__init__.py]: fixed

**Comment by jmchilton** [comment]
:+1:

**Comment by erasche** [comment]
@nsoranzo any other comments or can this be merged?

**Comment by nsoranzo** [comment]
Sorry Eric, I'm working on another open PR, will get to this soon.

**Comment by erasche** [comment]
Okay, no worries, just wanted to poke someone on bioblend about it. Thanks.

**Comment by erasche** [comment]
Okay...I hadn't intended to include these documentation changes in this PR, but due to carelessness I did. 

Please address questions in #118 before merging this PR so I can add those improvements in.

File: bioblend/galaxy/datasets/__init__.py [hunk 3]
```diff
@@ -62,6 +68,11 @@ def download_dataset(self, dataset_id, file_path=None, use_default_filename=True
         :param maxwait: Time (in seconds) to wait for dataset to complete.
                         If the dataset state is not complete within this time, a DatasetTimeoutException will be thrown.
 
+        :type file_ext: str
+        :param file_ext: Extension to request from Galaxy. Will default to
+                         dataset file_ext value. Provided for backwards
+                         compatability with HistoryClient.download_dataset()
```
Comment (nsoranzo) [inline bioblend/galaxy/datasets/__init__.py]: s/compatability/compatibility/

File: bioblend/galaxy/histories/__init__.py [hunk 4]
```diff
@@ -260,33 +337,25 @@ def download_dataset(self, history_id, dataset_id, file_path,
...
-        if use_default_filename:
-            file_local_path = os.path.join(file_path, meta['name'])
-        else:
-            file_local_path = file_path
 
-        with open(file_local_path, 'wb') as fp:
-            shutil.copyfileobj(req, fp)
+        Legacy use only, refer to
+        ``bioblend.galaxy.dataset.DatasetClient.download_dataset()`` for
+        available parameters
```
Comment (jmchilton) [inline bioblend/galaxy/histories/__init__.py]: Hmm... structurally I love that this is no longer duplicated and is now doing the right thing. Not sure about calling this endpoint for legacy use only tough. I may prefer the idea of having two logical ways to download datasets (a history content centric way and a dataset centric way - depending on your current flow).

Thoughts?

File: bioblend/galaxy/workflows/__init__.py [hunk 5]
```diff
@@ -137,21 +160,20 @@ def export_workflow_json(self, workflow_id):
 
     def export_workflow_to_local_path(self, workflow_id, file_local_path, use_default_filename=True):
         """
-        Exports a workflow in json format to a given local path.
+        Exports a workflow to a given local path.
```
Comment (nsoranzo) [inline bioblend/galaxy/workflows/__init__.py]: The file _is_ in JSON format, please revert this change.

File: bioblend/galaxy/workflows/__init__.py [hunk 6]
```diff
@@ -16,15 +16,25 @@ def get_workflows(self, workflow_id=None, name=None, deleted=False, published=Fa
...
+        :param name: Filter by name of workflow (incompatible with
+          ``workflow_id``). If multiple names match the given name, all the
+          workflows matching the argument will be returned.
+
+        :type deleted: bool
+        :param deleted: If set to ``True``, return workflows that have been
+          deleted.
+
+        :type published: bool
+        :param published: If set to ``True``, return published workflows.
```
Comment (nsoranzo) [inline bioblend/galaxy/workflows/__init__.py]: There has been a merge problem here, please remove the documentation for ```deleted``` and ```published``` above.

File: bioblend/galaxy/workflows/__init__.py (hunk 5 above)
Comment (erasche) [inline bioblend/galaxy/workflows/__init__.py]: Indeed it is

File: bioblend/galaxy/histories/__init__.py (hunk 4 above)
Comment (erasche) [inline bioblend/galaxy/histories/__init__.py]: Yay for less duplication.

I'm not sure I really understand the history-centric view of datasets, or rather why history was even a part of the dataset URL. I can understand wanting datasets from a history (and this convenience method for accessing it via the same class), but to my way of thinking accessing history information and accessing datasets wouldn't share the Galaxy API route.

File: bioblend/galaxy/workflows/__init__.py (hunk 6 above)
Comment (jmchilton) [inline bioblend/galaxy/workflows/__init__.py]: Done - https://github.com/galaxyproject/bioblend/commit/beac15289a1f5b2619d72cbdd0dbc02a91de2f53. Thanks for the catch.

File: bioblend/galaxy/histories/__init__.py [hunk 7]
```diff
@@ -167,16 +234,12 @@ def update_dataset(self, history_id, dataset_id, **kwds):
...
         :type history_id: str
         :param history_id: Encoded history ID
-        :type name: str
-        :param name: Replace history dataset name with the given string
-        :type annotation: str
-        :param annotation: Replace history dataset annotation with given string
-        :type deleted: bool
-        :param deleted: Mark or unmark history dataset as deleted
-        :type visible: bool
-        :param visible: Mark or unmark history dataset as visible
```
Comment (nsoranzo) [inline bioblend/galaxy/histories/__init__.py]: Why did you remove this part of the documentation? These are the parameters that may be specified in ```**kwds```.

Comment (erasche) [inline bioblend/galaxy/histories/__init__.py]: Hmmm. I think I removed them because they weren't explicit function parameters. That was wrong of me, I should've added them as extra arguments.
